### PR TITLE
windows compliance pack based on IAD SAMI

### PR DESCRIPTION
**Pull request by juju4**
Compliance check for windows based on https://github.com/iadgov/Splunk-Assessment-of-Mitigation-Implementations/raw/master/SAMI_Business_Logic.xlsx

### CODE REVIEW

File: packs/windows-iad-sami-compliance.conf [hunk 1]
```diff
@@ -0,0 +1,218 @@
+// inspired from https://github.com/iadgov/Splunk-Assessment-of-Mitigation-Implementations/raw/master/SAMI_Business_Logic.xlsx
+// default interval 1d=86400sec
+{
+  "platform": "windows",
+  "queries": {
+    // AE
+    "win_compliance_secureboot": {
```
Comment (muffins) [inline packs/windows-iad-sami-compliance.conf]: Perhaps we could just leave the pack name as `win_compliance`, and drop the `win_compliance_` prefix on all of these, as it seems a bit redundant.

File: packs/windows-iad-sami-compliance.conf [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+      "query" : "select * from registry where key like 'HKEY_LOCAL_MACHINE\\SYSTEM\\CurrentControlSet\\Control\\SecureBoot\\State\\UEFISecureBootEnabled'",
```
Comment (muffins) [inline packs/windows-iad-sami-compliance.conf]: Why the `like` here? Seems like you haev the precise registry path and you could get away with `key =` here and those below.

File: packs/windows-iad-sami-compliance.conf [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+      "interval": "86400",
+      "platform": "windows",
+      "snapshot": true,
+      "description" : "Secureboot state"
```
Comment (muffins) [inline packs/windows-iad-sami-compliance.conf]: This description is pretty terse, would you mind expanding this, and the others below? Perhaps "Whether Secure Boot is enabled on this system."?

File: packs/windows-iad-sami-compliance.conf [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+      "snapshot": true,
+      "platform": "windows"
+    },
+    "win_compliance_kernelsehopenabled": {
+      "query" : "select * from registry where key like 'HKEY_LOCAL_MACHINE\\SYSTEM\\CurrentControlSet\\Control\\Session Manager\\Kernel\\KernelSEHOPEnabled'",
+      "interval": "86400",
+      "snapshot": true,
+      "platform": "windows"
+    },
+    "win_compliance_enableCertPaddingCheck": {
```
Comment (muffins) [inline packs/windows-iad-sami-compliance.conf]: Can you stick with consistent casing? Either change this or modify those above.

File: packs/windows-iad-sami-compliance.conf [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+      "snapshot": true,
+      "platform": "windows"
+    },
+    "win_compliance_Rule": {
+      "query" : "select * from registry where key like 'HKEY_LOCAL_MACHINE\\SOFTWARE\\Policies\\Microsoft\\Windows\\SrpV2\\%\\%\\Value'",
+      "interval": "86400",
+      "snapshot": true,
+      "platform": "windows"
+    },
+    // HBSS
```
Comment (muffins) [inline packs/windows-iad-sami-compliance.conf]: Remove these three comments.

File: packs/windows-iad-sami-compliance.conf (hunk 1 above)
Comment (juju4) [reply packs/windows-iad-sami-compliance.conf]: fine to me

File: packs/windows-iad-sami-compliance.conf (hunk 2 above)
Comment (juju4) [reply packs/windows-iad-sami-compliance.conf]: I think I used like everywhere, exact path or %. can be updated

File: packs/windows-iad-sami-compliance.conf (hunk 3 above)
Comment (juju4) [reply packs/windows-iad-sami-compliance.conf]: yes. correct.

File: packs/windows-iad-sami-compliance.conf [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+      "description" : "Check ASLR configuration"
+    },
+    "KernelSehopEnabled": {
+      "query" : "select * from registry where key='HKEY_LOCAL_MACHINE\\SYSTEM\\CurrentControlSet\\Control\\Session Manager\\Kernel\\KernelSEHOPEnabled'",
+      "interval": "86400",
+      "snapshot": true,
+      "platform": "windows",
+      "description" : "Wether SEHOP is enabled"
+    },
+    "enableCertPaddingCheck": {
```
Comment (muffins) [inline packs/windows-iad-sami-compliance.conf]: ensure that youre casing is consistent throughout.

File: packs/windows-iad-sami-compliance.conf [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+      "snapshot": true,
+      "platform": "windows"
+    },
+    "appCompat": {
+      "query" : "select * from registry where key='HKEY_LOCAL_MACHINE\\SOFTWARE\\%Microsoft\\Windows NT\\CurrentVersion\\AppCompatFlags\\Layers'",
+      "interval": "86400",
+      "snapshot": true,
+      "platform": "windows"
+    },
+    "app_disabledExceptionChainValidation": {
```
Comment (muffins) [inline packs/windows-iad-sami-compliance.conf]: Similarly here, can you use consistent casing? either stick to upper CamlCase, lower camlCase, or snake_case?

File: packs/windows-iad-sami-compliance.conf [hunk 8]
```diff
@@ -0,0 +1,219 @@
+// inspired from https://github.com/iadgov/Splunk-Assessment-of-Mitigation-Implementations/raw/master/SAMI_Business_Logic.xlsx
+// default interval 1d=86400sec
+{
+  "platform": "windows",
+  "queries": {
+    // AE
```
Comment (muffins) [inline packs/windows-iad-sami-compliance.conf]: remove this comment.

**Comment by muffins** [comment]
ok to test

**Comment by osqueryer** [comment]
:-1: The commit bf833788b9add3cd9aa4056782731cf001b27db5 (Job results: [6618](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6618/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit bf833788b9add3cd9aa4056782731cf001b27db5 (Job results: [6619](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6619/)) failed one or more tests (**Linux**).

**Comment by muffins** [comment]
So, continuing on what I mentioned earlier, I think all of this information is really good to have, but all of these queries would be better categorized in other packs. The compliance concept is an interesting idea, and I think we'd be able to take the conversation of "should osquery have compliance packs that companies issue?", however for now I'd say all of these queries should be re-organized into the [`windows-hardening`](https://github.com/facebook/osquery/blob/master/packs/windows-hardening.conf#L3) and  [`it-compliance`](https://github.com/facebook/osquery/blob/master/packs/it-compliance.conf#L3) packs.

**Comment by juju4** [comment]
My personal take would be all in windows-hardening. IMHO compliance is a subset of hardening.
what is your preference?
Also, don't understand how linux build can fail on PR as it is windows related... platform is tagged.

**Comment by muffins** [comment]
@juju4 yeah go ahead and throw these in the windows hardening. As for the CI failures, that means there's something up with the JSON. We have integration tests that ensure the query packs are well formated JSON:
```
======================================================================
ERROR: test_pack_queries (__main__.ReleaseTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/jenkins/workspace/osqueryPullRequestLinux/TargetSystem/aws-ubuntu-14.04/tools/tests/test_base.py", line 549, in wrapper
    raise exceptions[0]
ValueError: No JSON object could be decoded

----------------------------------------------------------------------
Ran 3 tests in 6.031s

FAILED (errors=1)
  File "/var/jenkins/workspace/osqueryPullRequestLinux/TargetSystem/aws-ubuntu-14.04/tools/tests/test_release.py", line 44, in test_pack_queries
    packs[name] = json.loads(content)

  File "/var/jenkins/workspace/osqueryPullRequestLinux/TargetSystem/aws-ubuntu-14.04/tools/tests/test_release.py", line 44, in test_pack_queries
    packs[name] = json.loads(content)

  File "/var/jenkins/workspace/osqueryPullRequestLinux/TargetSystem/aws-ubuntu-14.04/tools/tests/test_release.py", line 44, in test_pack_queries
    packs[name] = json.loads(content)

Test (attempt 1) ReleaseTests::test_pack_queries failed: No JSON object could be decoded
Test (attempt 2) ReleaseTests::test_pack_queries failed: No JSON object could be decoded
Test (attempt 3) ReleaseTests::test_pack_queries failed: No JSON object could be decoded
```

**Pull request merged**